cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/DefaultSecureStoreService.java

```diff
@@ -0,0 +1,174 @@
+  private final SecureStore secureStore;
+  private final SecureStoreManager secureStoreManager;
+
+  @Inject
+  DefaultSecureStoreService(AuthorizerInstantiator authorizerInstantiator, AuthorizationEnforcer authorizationEnforcer,
+                            AuthenticationContext authenticationContext, SecureStore secureStore,
+                            SecureStoreManager secureStoreManager) {
+    this.authorizer = authorizerInstantiator.get();
+    this.authorizationEnforcer = authorizationEnforcer;
+    this.authenticationContext = authenticationContext;
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+  }
+
+  /**
+   * Lists all the secure keys in the given namespace that the user has access to.
+   * @param namespaceId Id of the namespace we want the key list for.
+   * @return A list of {@link SecureKeyListEntry} of all the keys visible to the user under the given namespace.
+   * @throws IOException If there was a problem getting the list from the underlying provider.
+   * @throws UnauthorizedException If we fail to create a filter for the current principal.
+   *
+   */
+  @Override
+  public List<SecureKeyListEntry> list(NamespaceId namespaceId) throws IOException, UnauthorizedException {
+    Principal principal = authenticationContext.getPrincipal();
+    List<SecureStoreMetadata> metadatas = secureStore.list(namespaceId.getNamespace());
+    List<SecureKeyListEntry> returnList = new ArrayList<>(metadatas.size());
+    final Predicate<EntityId> filter;
+    try {
+      filter = authorizationEnforcer.createFilter(principal);
+    } catch (Exception e) {
+      throw new UnauthorizedException(principal, Action.READ, namespaceId);
+    }
+
+    String namespace = namespaceId.getNamespace();
+    for (SecureStoreMetadata metadata : metadatas) {
+      String name = metadata.getName();
+      if (filter.apply(new SecureKeyId(namespace, name))) {
+        returnList.add(new SecureKeyListEntry(name, metadata.getDescription()));
+      }
+    }
+    return returnList;
+  }
+
+  /**
+   * Checks if the user has access to read the secure key and returns the data associated with the key if they do.
+   * @param secureKeyId Id of the key that the user is trying to read.
+   * @return Data associated with the key if the user has read access.
+   * @throws UnauthorizedException If we fail to create a filter for the current principal
+   * or the user does not have READ permissions on the secure key.
+   * @throws IOException If there was a problem getting the data from the underlying provider.
+   */
+  @Override
+  public SecureStoreData get(SecureKeyId secureKeyId) throws UnauthorizedException, IOException {
+    Principal principal = authenticationContext.getPrincipal();
+    final Predicate<EntityId> filter;
+    try {
+      filter = authorizationEnforcer.createFilter(principal);
+    } catch (Exception e) {
+      throw new UnauthorizedException(principal, Action.READ, secureKeyId);
```

Comment:
The authorization enforcer methods throw Exception. I thought you suggested separating them out.